autogluon_zeroshot/metrics/_roc_auc_cpp/cpp_auc.cpp

```diff
@@ -1,66 +1,131 @@
+#include <array>
 #include <vector>
-#include <iostream>
-#include <algorithm>
-#include <tuple>
-#include <type_traits>
-
-
-// Code based on https://github.com/diditforlulz273/fastauc , which is under MIT license
-
-// Fill the zipped vector with pairs consisting of the
-// corresponding elements of a and b. (This assumes
-// that the vectors have equal length)
-template <typename tuple_type> void zip(
-    const bool* a,
-    const float* b,
-    const size_t len,
-    std::vector<tuple_type> &zipped)
-    {
-        for(size_t i=0; i<len; ++i)
-        {
-            zipped.push_back(std::make_tuple(a[i], b[i]));
-        }
-    }
 
-double trapezoid_area(double x1, double x2, double y1, double y2) {
-  double dx = x2 - x1;
-  double dy = y2 - y1;
-  return dx * y1 + dy * dx / 2.0;
-}
 
-template <typename tuple_type> double auc_kernel(float* ts, bool* st, size_t len) {
-  // sort the data
-  // Zip the vectors together
-  std::vector<tuple_type> zipped;
-  zipped.reserve(len);
-  zip<tuple_type>(st, ts, len, zipped);
-
-  // Sort the vector of pairs
-  std::sort(std::begin(zipped), std::end(zipped),
-    [&](const auto& a, const auto& b)
-    {
-        return std::get<1>(a) > std::get<1>(b);
-    });
-
-  double fps = 0;
-  double tps = 0;
-  double last_counted_fps = 0;
-  double last_counted_tps = 0;
-  double auc = 0;
-  for (size_t i=0; i < zipped.size(); ++i) {
-    tps += std::get<0>(zipped[i]);
-    fps += (1 - std::get<0>(zipped[i]));
-    if ((i == zipped.size() - 1) || (std::get<1>(zipped[i+1]) != std::get<1>(zipped[i]))) {
-        auc += trapezoid_area(last_counted_fps, fps, last_counted_tps, tps);
-        last_counted_fps = fps;
-        last_counted_tps = tps;
-    }
+inline uint32_t mantissa(uint32_t a) { return a & 0x7FFFFF; }
+
+/*
+ * This function is a faster single-threaded implementation of the binary non-weighted ROC AUC metric.
+ * The primary speed-up come from a radix sort on the 23 bits of the mantissa of float values.
+ * This function makes several assumptions:
+ *   - y_true and y_pred are both of size len
+ *   - y_pred values are in range [0.0, 1.0)
```

Comment:
Why non-inclusive?

Reply:
It comes from the way floats are represented. But ultimately I want to sort numbers that come from a discrete range of floats that have a range of width 1, and [0.0, 1.0] is "technically" one bin larger than that. This discussion makes me think I need to add a corner case to make sure 1.0 is also handled correctly...

Reply:
Yes, additional testing shows this is indeed an issue. I will let you know when a fix is ready.